src/Interpreters/ExpressionActions.h

```diff
@@ -119,9 +118,29 @@ class ExpressionActions
     ExpressionActionsPtr clone() const;
 
 private:
+    struct Data
+    {
+        const Node * node = nullptr;
+        size_t num_created_children = 0;
+        std::vector<const Node *> parents;
+
+        ssize_t position = -1;
+        size_t num_created_parents = 0;
+        bool used_in_result = false;
+    };
+
     void checkLimits(const ColumnsWithTypeAndName & columns) const;
 
     void linearizeActions();
+
+    /// Enable lazy execution for short-circuit function arguments.
+    bool rewriteShortCircuitArguments(
+        const ActionsDAG::NodeRawConstPtrs & children, const std::unordered_map<const ActionsDAG::Node *, bool> & used_only_in_short_circuit_functions, bool force_enable_lazy_execution);
+
+    /// Find short-circuit functions in actions and enable lazy execution for actions that are used in their arguments.
+    void rewriteArgumentsForShortCircuitFunctions();
```

Comment:
Looks like this functions also don't need anything except DAG.
So, we also could make something like `rewriteArgumentsForShortCircuitFunctions` -> `static std::unordered_set<const ActionsDAG::Node *> getShortCircuitNodes(const ActionsDAG & actions_dag)` (and move to .cpp)